File: example/xor/xor.cc
// This example shows a small usage of the primitiv library with solving the XOR
// problem by 2-layer (input/hidden/output) perceptron.
//
// Compile:
// g++
//   -std=c++11
//   -I/path/to/primitiv/includes (typically -I../..)
//   -L/path/to/primitiv/libs     (typically -L../../build/primitiv)
//   xor.cc -lprimitiv

#include <iostream>

// "primitiv.h" can be used to include most of features in primitiv.
// Users also can include individual headers to suppress compilation cost.
#include <primitiv/primitiv.h>

// "primitiv_cuda.h" is required to use CUDA backend.
//#include <primitiv/primitiv_cuda.h>

// shortcuts
using std::cout;
using std::endl;
using std::vector;
using primitiv::initializers::Constant;
using primitiv::initializers::XavierUniform;
using primitiv::Device;
using primitiv::CPUDevice;
//using primitiv::CUDADevice;
using primitiv::Graph;
using primitiv::Node;
using primitiv::Parameter;
using primitiv::trainers::SGD;
using primitiv::Shape;
namespace F = primitiv::node_ops;

int main() {
  // Setups a computation backend.
  // The device object manages device-specific memories, and must be destroyed
  // after all other objects were gone.
  CPUDevice dev;
  Device::set_default_device(dev);

  // If you want to use CUDA, uncomment below line (and comment out above) with
  // a specific device (GPU) ID.
  //CUDADevice dev(0);

  // Parameters
  Parameter pw1("w1", {8, 2}, XavierUniform());
  Parameter pb1("b1", {8}, Constant(0.0f));
  Parameter pw2("w2", {1, 8}, XavierUniform());
  Parameter pb2("b2", {}, Constant(0.0f));

  // Trainer
  SGD trainer(0.1f);
  trainer.add_parameter(pw1);
  trainer.add_parameter(pb1);
  trainer.add_parameter(pw2);
  trainer.add_parameter(pb2);

  // Fixed input data
  const vector<float> input_data {
     1,  1,
     1, -1,
    -1,  1,
    -1, -1,
  };

  // Corresponding output data
  const vector<float> output_data {1, -1, -1, 1};

  // Training loop
  for (unsigned i = 0; i < 10; ++i) {
    Graph g;
    Graph::set_default_graph(g);
    
    // Builds a computation graph.
    Node x = F::input(Shape({2}, 4), input_data);
    Node w1 = F::input(pw1);
    Node b1 = F::input(pb1);
    Node w2 = F::input(pw2);
    Node b2 = F::input(pb2);
    Node h = F::tanh(F::matmul(w1, x) + b1);
    Node y = F::matmul(w2, h) + b2;

    // Calculates values.
    vector<float> y_val = g.forward(y).to_vector();
    cout << "epoch " << i << ":" << endl;
    for (unsigned j = 0; j < 4; ++j) {
      cout << "  [" << j << "]: " << y_val[j] << endl;
    }

    // Builds an additional computation graph for the mean squared loss.
    Node t = F::input(Shape({}, 4), output_data);
    Node diff = t - y;
    Node loss = F::batch::mean(diff * diff);
    
    // Calculates losses.
    // The forward() function performs over only additional paths.
    const float loss_val = g.forward(loss).to_vector()[0];
    cout << "  loss: " << loss_val << endl;

    // Resets cumulative gradients of all registered parameters.
    trainer.reset_gradients();

    // Backpropagation
    g.backward(loss);

    // Updates parameters.
    trainer.update();
  }

  return 0;
}

File: python-primitiv/examples/xor.py
#!/usr/bin/env python3

from primitiv import DefaultScope
from primitiv import Graph
from primitiv import Parameter
from primitiv import initializers as I
from primitiv import operators as F
from primitiv import trainers as T
from primitiv.devices import Naive

import numpy as np


def main():

    with DefaultScope(Naive()):
        pw1 = Parameter("w1", [8, 2], I.XavierUniform())
        pb1 = Parameter("b1", [8], I.Constant(0))
        pw2 = Parameter("w2", [1, 8], I.XavierUniform())
        pb2 = Parameter("b2", [], I.Constant(0))

        trainer = T.SGD(0.1)

        trainer.add_parameter(pw1)
        trainer.add_parameter(pb1)
        trainer.add_parameter(pw2)
        trainer.add_parameter(pb2)

        input_data = [
            np.array([ 1,  1], dtype=np.float32),  # Sample 1
            np.array([ 1, -1], dtype=np.float32),  # Sample 2
            np.array([-1,  1], dtype=np.float32),  # Sample 3
            np.array([-1, -1], dtype=np.float32),  # Sample 4
        ]

        output_data = [
             1,  # Label 1
            -1,  # Label 2
            -1,  # Label 3
             1,  # Label 4
        ]

        for i in range(100):
            g = Graph()
            with DefaultScope(g):
                # Builds a computation graph.
                x = F.input(input_data)
                w1 = F.parameter(pw1)
                b1 = F.parameter(pb1)
                w2 = F.parameter(pw2)
                b2 = F.parameter(pb2)
                h = F.tanh(F.matmul(w1, x) + b1)
                y = F.matmul(w2, h) + b2

                # Calculates values.
                y_val = g.forward(y).to_list()
                print("epoch ", i, ":")
                for j in range(4):
                    print("  [", j, "]: ", y_val[j])
                    t = F.input(output_data)
                diff = t - y
                loss = F.batch.mean(diff * diff)
                loss_val = g.forward(loss).to_list()[0]
                print("  loss: ", loss_val)
                trainer.reset_gradients()
                g.backward(loss)
                trainer.update()


if __name__ == "__main__":
    main()
